src/main/java/randoop/generation/ForwardGenerator.java

```
@@ -981,4 +1031,45 @@ public String toString() {
                 "runtimePrimitivesSeen: " + runtimePrimitivesSeen.size()))
         + ")";
   }
+
+  /**
+   * Selects one sequence from {@code candidates} using TF-IDF weights computed for the scope
+   * associated with {@code type}. Returns {@code null} if {@code candidates} is empty or the scope
+   * has no literal statistics.
+   *
+   * @param candidates candidate sequences that produce values of the needed type
+   * @param type the type whose scope determines the TF-IDF statistics
+   * @param scopeToLiteralStatistics provider of literal statistics and scope resolution
+   * @return the TF-IDF–weighted choice, or {@code null} if unavailable
+   */
+  private @Nullable Sequence selectTfidfSequence(
+      SIList<Sequence> candidates,
+      ClassOrInterfaceType type,
+      ScopeToLiteralStatistics scopeToLiteralStatistics) {
+
+    if (candidates.isEmpty()) {
+      return null;
+    }
+
+    // literalStats contains all literal-bearing sequences known for the resolved scope,
+    // regardless of the specific needed input type. The candidates passed here are a typed
+    // subset taken from those statistics.
+    LiteralStatistics literalStats = scopeToLiteralStatistics.getLiteralStatistics(type);
+
+    if (literalStats.isEmpty()) {
+      return null;
```

I think this should throw an exception rather than returning `null`.

---

I'm confused by this suggestion. Consider the following example (this is the way I understand it):

Scope: CLASS mode
Operation: Testing a method in class "MyUtils"
Declaring type: MyUtils

Scenario 1: MyUtils has bytecode literals
  1.1: ClassLiteralExtractor.visitBefore(MyUtils.class) runs
  1.2: ClassFileConstants.getConstants("MyUtils") extracts [1, "hello", 3.14, "world"]
  1.3: incrementNumUses() populates literalUsesByType with these constants
  1.4: literalStats.isEmpty() == false
  1.5: TF-IDF selection proceeds normally

Scenario 2: MyUtils has NO bytecode literals (e.g., all constants inline, no const pool entries)
  2.1: ClassLiteralExtractor.visitBefore(MyUtils.class) runs but finds nothing
  2.2: No calls to incrementNumUses() or recordSequencesInClass()
  2.3: literalUsesByType remains empty
  2.4: literalStats.isEmpty() == true
  2.5: selectTfidfSequence() hits the check:
      if (literalStats.isEmpty()) {
          return null;  // ← No literals available
      }
  2.6: Caller (selectInputs) falls back to non-TF-IDF literal selection